### PR TITLE
Added Controlled Adams Bashforth Moulton Stepper

### PR DESCRIPTION
**Pull request by vhartman**
Implemented a controlled adams bashforth moulton Stepper requested by Issue #175.
The stepper is based on the divided difference method, which is used to calculate the coefficients for the prediction and correction step. The order can be chosen arbitrarily large. I added the stepper to the integration-test as well.

The currently implemented stepsize controller is described in [0]. 

[0] http://www.maths.lth.se/na/staff/gustaf/filtern.pdf

### CODE REVIEW

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+		{
+			do_step_impl(system, m_coeff, in, t, out, dt, xerr);
+
+			system(out, m_dxdt.m_v, t+dt);
+			m_coeff.step(m_dxdt.m_v, t+dt);
+			m_coeff.confirm();
+		};
+
+		template<class System>
+		void do_step_impl(System system, coeff_type coeff, const state_type & in, time_type &t, state_type & out, time_type &dt, state_type &xerr)
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp]: _impl should be private

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+			m_dxdt_resizer.adjust_size( in , detail::bind( &stepper_type::template resize_dxdt_impl< state_type > , detail::ref( *this ) , detail::_1 ) );
+
+			system(in, m_dxdt.m_v, t);
+			m_coeff.step(m_dxdt.m_v, t);
+			m_coeff.confirm();
+
+			do_step_impl(m_coeff, in, t, out, dt);
+		};
+
+		void do_step_impl(coeff_type & coeff, const state_type & in, time_type t, state_type & out, time_type dt)
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: Why do we need to pass coeff here? Can't it just use m_coeff directly?

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+class Algebra = typename algebra_dispatcher< State >::algebra_type,
+class Operations = typename operations_dispatcher< State >::operations_type ,
+class Resizer = initially_resizer
+>
+class adaptive_adams_bashforth: public algebra_stepper_base< Algebra , Operations >
+{
+	public:
+		static const size_t steps = Steps;
+		typedef unsigned short order_type;
+		static const order_type order_value = steps;
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: Is this really true? If I understand the implementation below correctly it calls the rhs function exactly once? I find it hard to believe this can result in a higher order.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp (hunk 1 above)
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp]: remove coeff from parameters (if possible?)

File: include/boost/numeric/odeint/stepper/adaptive_adams_moulton.hpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		void do_step(coeff_type & coeff, const state_type & in,
+					time_type t, state_type & out, time_type &dt)
+		{
+			coeff.poly.add_root(0);
+			out = in;
+
+			// integrating
+			for(size_t i=0; i<coeff.m_effective_order; ++i)
+			{
+				time_type c = ((i!=coeff.m_effective_order-1)?coeff.m_c[i]:coeff.poly.evaluate_integrated(dt));
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_moulton.hpp]: can you add some white spaces for readability here?

File: include/boost/numeric/odeint/stepper/adaptive_adams_moulton.hpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		adaptive_adams_moulton( const algebra_type &algebra = algebra_type())
+		:algebra_stepper_base_type( algebra )
+		{};
+
+		order_type order() const
+		{
+			return order_value;
+		};
+
+		void do_step(coeff_type & coeff, const state_type & in,
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_moulton.hpp]: This class is not really meant to be used on its own I assume?

File: include/boost/numeric/odeint/stepper/detail/adaptive_adams_coefficients.hpp [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		typedef Resizer resizer_type;
+
+		typedef adaptive_adams_coefficients<Steps, Deriv, Time, Algebra, Operations, Resizer> aac_type;
+		typedef detail::Polynomial<steps+2, time_type> poly_type;
+
+		adaptive_adams_coefficients(const algebra_type &algebra = algebra_type() )
+		:poly(), m_effective_order(1), m_resizer(), m_algebra(algebra)
+		{};
+
+		/*aac_type &operator=(const aac_type & rhs)
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/detail/adaptive_adams_coefficients.hpp]: please remove commented code

File: include/boost/numeric/odeint/stepper/detail/pid_step_adjuster_coefficients.hpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+	H312PID,
+	H0321,
+	H321
+};
+
+template<int Type>
+class pid_step_adjuster_coefficients;
+
+template<>
+class pid_step_adjuster_coefficients<BASIC> : public boost::array<double, 5>
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/detail/pid_step_adjuster_coefficients.hpp]: the floating point type should be passed on from the stepper's value_type. But I think this can be done in a later PR, this is quite big already.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp (hunk 3 above)
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: I am not sure what you are saying. The standard adams-bashforth only calls the rhs once as well, and the order is exactly the same as it is here.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+		};
+
+		template<class System>
+		void do_step(System system, const state_type & in, time_type t, state_type & out, time_type dt)
+		{
+			m_dxdt_resizer.adjust_size( in , detail::bind( &stepper_type::template resize_dxdt_impl< state_type > , detail::ref( *this ) , detail::_1 ) );
+
+			system(in, m_dxdt.m_v, t);
+			m_coeff.step(m_dxdt.m_v, t);
+			m_coeff.confirm();
```
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: The problem is that do_step_impl is called from adaptive_adams_bashforth_moulton with coeff. However, when the step is not accepted, the coefficients should not be confirmed, and therefore the confirmation is not done in do_step_impl. This might not be the best solution, but I was not able to find something better.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp (hunk 2 above)
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: It can be called from adaptive_adams_bashforth, which is why it passes coeff

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp (hunk 1 above)
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp]: Called from controlled_adams_bashforth_moulton for the same reason as above (coeff should not be confirmed if the step is rejected)

File: include/boost/numeric/odeint/stepper/adaptive_adams_moulton.hpp (hunk 5 above)
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_moulton.hpp]: Yes, it is equivalent to adams_moulton.

**Comment by vhartman** [comment]
Thank you for the comments! I added the testcases for the numeric precision. There seems to be a problem with the higher order adaptive_adams_bashforth_moulton (The magnitude of the error is correct, but still slightly over the upper bound).

I am currently looking into the implementation of a slightly different formulation which allows to avoid new calculations of the coefficients if the stepsize is the same.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp (hunk 3 above)
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: Right, sorry. I forgot that these are multi-step methods... Nevermind.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp (hunk 8 above)
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: Right, I see. Would you expect someone to use adaptive_adams_bashforth by itself. If not, this could be moved int detail and not exposed to the user at all. Otherwise I think we can leave it as is, also fine with me.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp [hunk 9]
```diff
@@ -89,6 +89,18 @@ class adaptive_adams_bashforth: public algebra_stepper_base< Algebra , Operation
...
     };
 
+    template<class System>
+    void initialize(System system, state_type &inOut, time_type &t, time_type dt)
+    {
+        m_coeff.reset();
+
+        for(size_t i=0; i<steps; ++i)
+        {
+            do_step(system, inOut, t, dt);
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: Out of curiosity: would it be possible to use a smaller stepsize of dt/steps here during initialization? Does this make sense? I would ensure that initialization if also of higher order, maybe? However, this can be done later (if at all).

**Comment by mariomulansky** [comment]
Looks good so far. I can merge if you want and you can add improvements in a following PR. Or we wait and you check the other formulation first, however you want. As for the numerical check: The upper bound (I think its called `f` in the code) is only hand-wavy. There is an a priori unknown pre-factor that I adjusted such that it passes the tests. If the order is right you might just have to increase f slightly to have things pass.

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp (hunk 9 above)
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: Would be possible, yeah. Another possibility would be to make use of an initializing stepper (same as done in adams_bashforth)

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp (hunk 8 above)
Comment (vhartman) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth.hpp]: I don't think there is any reason to use adaptive_adams_bashforth over adams_bashforth. But in that case, it would probably make more sense to remove adaptive_adams_bashforth and adaptive_adams_moulton altogether and move the complete logic into adaptive_adams_bashforth_moulton.

**Comment by vhartman** [comment]
I would prefer to finish the other formulation first, and push it here. It might take another week though.
Thank you very much!

**Comment by vhartman** [comment]
Hey! Thank you for the patience, the stepper should be pretty much done now. There are a few things that I am not sure if they could be designed better, and a few things that could be improved (e.g. pre-computation of the coefficients, additional controllers), but it should be fine for now.

Could you review it again, and if you don't want me anything to change, merge it? I would then start the documentation, and open another pull request for that.

**Comment by vhartman** [comment]
Alright, now it should be actually done. I just added proper treatment of the absolute and relative error (in the same way as it is done in controlled_runge_kutta).

File: include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        m_dxdt_resizer.adjust_size( in , detail::bind( &stepper_type::template resize_dxdt_impl< state_type > , detail::ref( *this ) , detail::_1 ) );
+
+        m_coeff.predict(t, dt);
+        if (eO == 1)
+        {
+            system(in, m_dxdt.m_v, t);
+            m_coeff.do_step(m_dxdt.m_v, 1);
+        }
+
+        out = in;
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/adaptive_adams_bashforth_moulton.hpp]: please use odeint::copy

File: include/boost/numeric/odeint/stepper/detail/adaptive_adams_coefficients.hpp [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+        beta[0][0] = 1;
+        beta[1][0] = 1;
+    };
+
+    void predict(time_type t, time_type dt)
+    {
+        m_time_storage[0] = t;
+
+        if (fabs(m_time_storage[0] - m_time_storage[1] - dt) > 1e-16 || m_eo >= m_ns)
```
Comment (mariomulansky) [inline include/boost/numeric/odeint/stepper/detail/adaptive_adams_coefficients.hpp]: Does this work with negative time steps? Odeint supports integrating "backwards", e.g. t0=10, t1=0, dt=-0.1. Does the controlled Adams-Bashforth also allows for that? Tests with negative timesteps are here: https://github.com/headmyshoulder/odeint-v2/blob/master/test/integrate.cpp for example.
These helpers might be useful: https://github.com/headmyshoulder/odeint-v2/blob/master/include/boost/numeric/odeint/util/detail/less_with_sign.hpp

Comment (vhartman) [inline include/boost/numeric/odeint/stepper/detail/adaptive_adams_coefficients.hpp]: I added adaptive_adams_bashforth_moulton and controlled_adams_bashforth_moulton to the tests in integrate.cpp, so it should work.

**Comment by mariomulansky** [comment]
great work, thanks a lot!

**Comment by vhartman** [comment]
Thanks for merging!
I have found a bug in the initialization, that somehow slipped through the numerical unit test (apparently nan is larger than nan).

Should I open another pullrequest?

**Comment by mariomulansky** [comment]
sure, just open another PR.
